copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopybookSource.java

```diff
@@ -99,77 +119,195 @@ public void transform(KeyValue<LongWritable, Map<String, AbstractFieldValue>> in
                         Emitter<StructuredRecord> emitter)
     throws Exception {
     Map<String, AbstractFieldValue> value = input.getValue();
-    if (outputSchema == null) {
-      outputSchema = getSchema(value.keySet());
-    }
     StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
     for (Schema.Field field : outputSchema.getFields()) {
       String fieldName = field.getName();
       if (value.containsKey(fieldName)) {
-        builder.set(fieldName, parseValue(field, value.get(fieldName)));
+        builder.set(fieldName, getFieldValue(value.get(fieldName)));
       }
     }
     emitter.emit(builder.build());
   }
 
   /**
-   * Convert each field to data type as specified by user in the output schema
+   * Get the output schema from the COBOL copybook contents specified by the user.
    *
-   * @param field Schema.Field object defining the field name and properties
-   * @param value value of field to be converted to datatype specified by user
-   * @return value casted to the required data type
-   * @throws IOException
+   * @return outputSchema
    */
-  private Object parseValue(Schema.Field field, @Nullable AbstractFieldValue value) throws IOException {
-    String fieldName = field.getName();
-    Schema fieldSchema = field.getSchema();
-    if (value == null) {
-      if (!fieldSchema.isNullable()) {
-        throw new IllegalArgumentException("NULL value found for non-nullable field : " + fieldName);
+  private Schema getOutputSchema() {
+
+    InputStream inputStream = null;
+    ExternalRecord externalRecord = null;
+    List<Schema.Field> fields = Lists.newArrayList();
+    try {
+      inputStream = IOUtils.toInputStream(config.copybookContents, "UTF-8");
+      BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
+      externalRecord = CopybookIOUtils.getExternalRecord(bufferedInputStream);
+
+      String fieldName;
+      for (ExternalField field : externalRecord.getRecordFields()) {
+        fieldName = field.getName();
+        if ((fieldsToDrop.size() > 0 && !fieldsToDrop.contains(fieldName)) || fieldsToDrop.size() == 0) {
```

Comment:
can be simplified to `if (fieldsToDrop.size() == 0 || !fieldsToDrop.contains(fieldName))`